pkg/job/job.go

```diff
@@ -0,0 +1,217 @@
+type Job interface {
+	// Start the job in the background
+	Start() error
+	// Stop the job
+	Stop(force bool) error
+	// Query the current status of the job
+	Status() *proto.JobStatus
+}
+
+// TODO Handle QUEUING state when job queue and scheduler is implemented
+// For now, the job will be start immediately after created
+// Process waiting to be executed
+
+// Impl is the implementation of the Job interface
+type Impl struct {
+	id         uuid.UUID
+	cmd        *exec.Cmd
+	logger     *log.Logger
+	owner      string
+	state      proto.ProcessState
+	stateMutex sync.RWMutex
+	exitChan   chan bool
+}
+
+// newJob creates a new job, associated with a unique ID and a Command, and the owner who requests its
+func newJob(ID uuid.UUID, cmd *exec.Cmd, logger *log.Logger, owner string) (Job, error) {
+	return &Impl{
+		id:       ID,
+		cmd:      cmd,
+		logger:   logger,
+		owner:    owner,
+		state:    proto.ProcessState_RUNNING,
+		exitChan: make(chan bool, 1),
+	}, nil
+}
+
+// Start runs job in the background
+func (j *Impl) Start() error {
+	// IO Mapping
+	stdoutLog := j.logger.Entry.WithField("source", "stdout").Writer()
+	stderrLog := j.logger.Entry.WithField("source", "stderr").Writer()
+
+	j.cmd.Stdout = stdoutLog
+	j.cmd.Stderr = stderrLog
+	j.cmd.Stdin = nil
+
+	j.cmd.SysProcAttr = configNameSpace()
+
+	logrus.Infof("Starting job %s", j.id)
+
+	err := j.cmd.Start()
+	if err != nil {
+		j.changeState(proto.ProcessState_STOPPED)
+		return err
+	}
+
+	go func() {
+		if err := j.cmd.Wait(); err != nil {
+			j.logger.Entry.Infof("Job %s terminated, reason %s", j.id, err)
+		}
+
+		// Close logger
+		if err := stdoutLog.Close(); err != nil {
+			logrus.Errorln(err)
+		}
+
+		if err := stderrLog.Close(); err != nil {
+			logrus.Errorln(err)
+		}
+
+		j.logger.Entry.Infof("Exiting job %s", j.id)
+
+		if err := j.logger.Close(); err != nil {
+			logrus.Errorln(err)
+		}
+
+		j.changeState(proto.ProcessState_EXITED)
+
+		// Inform stop caller to finish.
+		// The channel buffer size is 1, so if this job exit naturally, it won't be blocked
+		j.exitChan <- true
+		close(j.exitChan)
+	}()
+
+	return nil
+}
+
+// Stop terminates a job. If the force flag is set to true, it will using SIGKILL to terminate the process, otherwise it will be SIGTERM
+func (j *Impl) Stop(force bool) error {
+	if j.getState() != proto.ProcessState_RUNNING || j.cmd.Process == nil {
+		return ErrNotRunning
+	}
+
+	// Use SIGKILL to force the process to stop immediately, otherwise we will use SIGTERM
+	if force {
+		if err := j.cmd.Process.Signal(os.Kill); err != nil {
+			return err
+		}
+	} else {
+		if err := j.cmd.Process.Signal(syscall.SIGTERM); err != nil {
+			return err
+		}
+	}
+
+	// Wait for the job to exit with timeout of 1s
+	select {
+	case _, ok := <-j.exitChan:
+		if !ok {
+			// exit channel is already closed, which means the process is already exited
+			return ErrNotRunning
+		}
+	case <-time.After(time.Second):
+		return ErrTimeout
```

Comment:
Hi, This is the case where a signal cannot stop a job or it might take too long. The timeout will prevent deadlock in the stop request and return with a grpc DeadlineExceeded code. After this, if the process can react to the signal it will stop, otherwise user has to send another stop request with the force option